src/Plutarch/Extra/AssetClass.hs

```diff
@@ -1,97 +1,361 @@
+import qualified GHC.Generics as GHC
+
+--------------------------------------------------------------------------------
+
 import Plutarch.Api.V1 (
     PCurrencySymbol,
     PTokenName,
-    PValue (PValue),
- )
-import Plutarch.Api.V1.AssocMap (PMap (PMap))
-import Plutarch.Api.V2 (
-    AmountGuarantees,
-    KeyGuarantees,
  )
-import Plutarch.DataRepr (PDataFields)
-import Plutarch.Extra.TermCont (pletC, pmatchC)
+import Plutarch.Unsafe (punsafeCoerce)
+
+--------------------------------------------------------------------------------
+
+import Plutarch.Orphans ()
+
+--------------------------------------------------------------------------------
+
+import qualified PlutusLedgerApi.V1.Value as Plutus
+
+--------------------------------------------------------------------------------
+-- AssetClass & Variants
+
+{- | Version of Plutus AssetClass, with currency unit tag
+ @since 3.8.0
+-}
+data AssetClass (tag :: Symbol) = AssetClass
+    { symbol :: Plutus.CurrencySymbol
+    , name :: Plutus.TokenName
+    }
+    deriving stock
+        ( -- | @since 3.8.0
+          Eq
+        , -- | @since 3.8.0
+          Ord
+        , -- | @since 3.8.0
+          Show
+        , -- | @since 3.8.0
+          GHC.Generic
+        )
+    deriving anyclass
+        ( -- | @since 3.8.0
+          ToJSON
+        , -- | @since 3.8.0
+          FromJSON
+        , -- | @since 3.8.0
+          FromJSONKey
+        , -- | @since 3.8.0
+          ToJSONKey
+        )
 
--- | @since 1.0.0
-newtype PAssetClass (s :: S)
-    = PAssetClass
+{- | Scott-encoded AssetClass
+ @since 3.8.0
+-}
+data PAssetClass (unit :: Symbol) (s :: S) = PAssetClass
+    { psymbol :: Term s (PAsData PCurrencySymbol)
+    , pname :: Term s (PAsData PTokenName)
+    }
+    deriving stock
+        ( -- | @since 3.8.0
+          GHC.Generic
+        )
+    deriving anyclass
+        ( -- | @since 3.8.0
+          PEq
+        , -- | @since 3.8.0
+          PlutusType
+        )
+
+-- | @since 3.8.0
+instance DerivePlutusType (PAssetClass unit) where
+    type DPTStrat _ = PlutusTypeScott
+
+-- | @since 3.8.0
+pconstantCls ::
+    forall (unit :: Symbol) (s :: S).
+    AssetClass unit ->
+    Term s (PAssetClass unit)
+pconstantCls (AssetClass sym tk) =
+    pcon $
+        PAssetClass (pconstantData sym) (pconstantData tk)
+
+{- | Coerce the unit tag of a PAssetClass
+ | @since 3.8.0
+-}
+pcoerceCls ::
+    forall (b :: Symbol) (a :: Symbol) (s :: S).
+    Term s (PAssetClass a) ->
+    Term s (PAssetClass b)
+pcoerceCls = punsafeCoerce
+
+{- | Construct a PAssetClass with empty TokenName
+ | @since 3.8.0
+-}
+psymbolAssetClass ::
+    forall (unit :: Symbol) (s :: S).
+    Term s (PAsData PCurrencySymbol) ->
+    PAssetClass unit s
+psymbolAssetClass sym = PAssetClass sym emptyTokenNameData
+
+--------------------------------------------------------------------------------
+-- Simple Helpers
+
+{- | Check whether an AssetClass is the Ada Class
```

Comment:
Same here. In fact, it's actually quite important to correctly-link Haddocks here, since `AssetClass` exists in `plutus-tx` as well.